### PR TITLE
style: make all mutable function arguments "None" and init in function body

### PR DESCRIPTION
**Pull request by jfyuen**
Mutable arguments were used as parameter to http requests. So they are "side effect" free wrt the runtime, but it is still better to avoid mutability

### CODE REVIEW

File: dataikuapi/dssclient.py [hunk 1]
```diff
@@ -966,7 +978,7 @@ def execute(self, settings = {}):
         @warning: You must check the 'success' flag
         """
         # Empty JSON dicts can't be parsed properly
-        if settings == {}:
-            settings["_"] = "_"
+        if settings is None:
```
Comment (nicolasservel) [inline dataikuapi/dssclient.py]: In that case, if the user actually passes `sesttings={}`, the behaviour changes

Comment (jfyuen) [reply dataikuapi/dssclient.py]: Nice catch!

**Comment by cstenac** [comment]
I don't understand what this fixes

**Comment by instanceofme** [comment]
No effective current bug but implements a best practice against Python's surprising behavior of [mutable default arguments](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments):

```python
>>> def func(arg=[]):
...   arg.append(1)
...   return len(arg)
...
>>> func()
1
>>> func()
2
>>> func()
3
```

File: dataikuapi/dss/project.py [hunk 2]
```diff
@@ -676,18 +697,20 @@ def get_scenario(self, scenario_id):
...
         """
         Create a new scenario in the project, and return a handle to interact with it
 
         :param str scenario_name: The name for the new scenario. This does not need to be unique
                                 (although this is strongly recommended)
         :param str type: The type of the scenario. MUst be one of 'step_based' or 'custom_python'
-        :param object definition: the JSON definition of the scenario. Use ``get_definition(with_status=False)`` on an
-                existing ``DSSScenario`` object in order to get a sample definition object
+        :param dict definition: the JSON definition of the scenario. Use ``get_definition(with_status=False)`` on an
+                existing ``DSSScenario`` object in order to get a sample definition object (defaults to `{}`)
```
Comment (nicolasservel) [inline dataikuapi/dss/project.py]:
```suggestion
                existing ``DSSScenario`` object in order to get a sample definition object (defaults to `{'params': {}}`)
```

File: dataikuapi/dss/macro.py [hunk 3]
```diff
@@ -33,16 +33,20 @@ def get_definition(self):
...
+    def run(self, params=None, admin_params=None, wait=True):
         """
         Run the macro from the project
 
-        :param params: parameters to the macro run
-        :param admin_params: admin parameters to the macro run (if the authentication of
-                             the api client does not cover admin rights, they are ignored)
+        :param dict params: parameters to the macro run (defaults to `{}`)
+        :param dict admin_params: admin parameters to the macro run (if the authentication of
+                             the api client does not cover admin rights, they are ignored, default to `{}`)
```
Comment (nicolasservel) [inline dataikuapi/dss/macro.py]:
```suggestion
                             the api client does not cover admin rights, they are ignored, defaults to `{}`)
```

File: dataikuapi/dssclient.py [hunk 4]
```diff
@@ -780,10 +790,15 @@ def get_apideployer(self):
     # Data Catalog
     ########################################################
 
-    def catalog_index_connections(self, connection_names=[], all_connections=False, indexing_mode="FULL"):
+    def catalog_index_connections(self, connection_names=None, all_connections=False, indexing_mode="FULL"):
         """
         Triggers an indexing of multiple connections in the data catalog
+
+        :param list connection_names: list of connections to index, ignored if `all_connections=True` (defaults to `[]`)
+        :param bool all_connections: index all connections, (defaults to `False`)
```
Comment (nicolasservel) [inline dataikuapi/dssclient.py]:
```suggestion
        :param bool all_connections: index all connections (defaults to `False`)
```

**Comment by nicolasservel** [comment]
Run locally integration tests (`scenario/ml/prediction/publicapi` and `scenario/misc/publicapi`) and they are all passing, mergin

**Pull request merged**